In [15]:
%load_ext autoreload
%autoreload 2

In [44]:
import pandas as pd
import re

from utils import get_data, Config

## Load Data + Preprocessing

In [321]:
# Load all data
initial_data = """
SELECT "Result".*, 
"Response".raw_header, "Response".status_code, "Response".label, "Response".resp_type,
"Browser".name, "Browser".version, "Browser".headless_mode, "Browser".os, "Browser".automation_mode, "Browser".add_info
FROM "Result"
JOIN "Response" ON "Result".response_id = "Response".id JOIN "Browser" ON "Result".browser_id = "Browser".id
WHERE "Response".created_at <= DATE('2023-11-21') and "Browser".name != 'Unknown';
"""
df = get_data(Config(), initial_data)

Connecting to the PostgreSQL database...
Connection successful


In [322]:
# Extra columns
df["outcome_str"] = df["outcome_value"].fillna("None").astype(str)

# Unify outcomes that are the same
# Firefox: {'error': 'object "TypeError: NetworkError when attempting to fetch resource."', 'headers': ''} Chromium: {'error': 'object "TypeError: Failed to fetch"', 'headers': ''}
# Safari: {'error': 'object "TypeError: Load failed"', 'headers': ''}
df["outcome_str"] = df["outcome_str"].replace("{'error': 'object \"TypeError: NetworkError when attempting to fetch resource.\"', 'headers': ''}", "{'error': 'object \"TypeError: Failed to fetch\"', 'headers': ''}")
df["outcome_str"] = df["outcome_str"].replace("{'error': 'object \"TypeError: Load failed\"', 'headers': ''}", "{'error': 'object \"TypeError: Failed to fetch\"', 'headers': ''}")

# For document referrer we do not want to know the exact resp_id and count
# Maybe we even only want to know whether it is a origin or the full URl?
df['outcome_str'] = df['outcome_str'].replace(r'resp_id=\d+', 'resp_id=<resp_id>', regex=True)
df['outcome_str'] = df['outcome_str'].replace(r'count=\d+', 'count=<count>', regex=True)

In [323]:
df["outcome_str"].unique()

array(["{'error': 'null', 'headers': 'access-control-allow-orgin,content-length,date,server,test,'}",
       '{\'error\': \'object "TypeError: Failed to fetch"\', \'headers\': \'\'}',
       "{'error': 'null', 'headers': 'access-control-allow-credentials,content-length,date,server,test,'}",
       "{'error': 'null', 'headers': 'access-control-allow-credentials,access-control-allow-orgin,content-length,date,server,test,'}",
       'message send',
       "{'error': 'null', 'headers': 'access-control-allow-credentials,access-control-allow-headers,access-control-allow-methods,access-control-allow-orgin,access-control-expose-headers,content-length,date,server,test,'}",
       "{'error': 'null', 'headers': 'cache-control,content-type,'}",
       'message timeout', 'fullscreenEnabled: true',
       'fullscreenEnabled: false', 'window.originAgentCluster: false',
       'window.originAgentCluster: true', "{'requestStart != 0': True}",
       "{'requestStart != 0': False}",
       'document.refe

## Overview

In [324]:
display(df.head(2))
display(df.describe())
display(df.columns)

id    outcome_type                                      outcome_value  \
0   1  <class 'dict'>  {'error': 'null', 'headers': 'access-control-a...   
1   2  <class 'dict'>  {'error': 'null', 'headers': 'access-control-a...   

    test_name  test_status test_message test_stack org_scheme  \
0   fetch_GET            0         None       None       http   
1  fetch_TEST            0         None       None       http   

                      org_host resp_scheme  ... status_code label  resp_type  \
0  sub.headers.websec.saarland        http  ...         200  CORS      basic   
1  sub.headers.websec.saarland        http  ...         200  CORS      basic   

     name  version headless_mode            os automation_mode add_info  \
0  chrome      119  headless-new  Ubuntu 22.04        selenium            
1  chrome      119  headless-new  Ubuntu 22.04        selenium            

                                         outcome_str  
0  {'error': 'null', 'headers': 'access-control-a...  
1  {'error': 'null', 'headers': 'access-control-a...  

[2 rows x 30 columns]

id    test_status     browser_id  testcase_id  \
count  198664.000000  198664.000000  198664.000000     198664.0   
mean    99332.500000       0.000705      55.105263          1.0   
min         1.000000       0.000000       5.000000          1.0   
25%     49666.750000       0.000000      13.000000          1.0   
50%     99332.500000       0.000000      60.000000          1.0   
75%    148998.250000       0.000000      75.000000          1.0   
max    198664.000000       2.000000     116.000000          1.0   
std     57349.501277       0.037536      40.144439          0.0   

         response_id                     created_at  \
count  198664.000000                         198664   
mean       78.027927  2023-11-21 12:45:18.583738112   
min         3.000000     2023-11-21 10:34:47.097638   
25%        23.000000  2023-11-21 11:38:47.412738560   
50%        46.000000  2023-11-21 12:46:22.775672064   
75%       151.000000  2023-11-21 13:50:22.174202368   
max       217.000000     2023-11-21 15:13:12.751466   
std        66.518630                            NaN   

                          updated_at    status_code  
count                         198664  198664.000000  
mean   2023-11-21 12:45:18.583738368     215.998470  
min       2023-11-21 10:34:47.097641     200.000000  
25%    2023-11-21 11:38:47.412740352     200.000000  
50%    2023-11-21 12:46:22.775673088     200.000000  
75%    2023-11-21 13:50:22.174202368     200.000000  
max       2023-11-21 15:13:12.751467     302.000000  
std                              NaN      37.093123

Index(['id', 'outcome_type', 'outcome_value', 'test_name', 'test_status',
       'test_message', 'test_stack', 'org_scheme', 'org_host', 'resp_scheme',
       'resp_host', 'relation_info', 'browser_id', 'testcase_id',
       'response_id', 'status', 'full_url', 'created_at', 'updated_at',
       'raw_header', 'status_code', 'label', 'resp_type', 'name', 'version',
       'headless_mode', 'os', 'automation_mode', 'add_info', 'outcome_str'],
      dtype='object')

In [325]:
# Collected results per browser
df.groupby(["name", "version", "os", "headless_mode", "automation_mode"])["id"].count().sort_values(ascending=False).to_frame()

id
name    version                   os           headless_mode automation_mode       
brave   1.59.120 (118.0.5993.88)  Ubuntu 22.04 headless-new  selenium         10456
                                               xvfb          selenium         10456
        1.60.114 (119.0.6045.124) Ubuntu 22.04 headless-new  selenium         10456
                                               xvfb          selenium         10456
        1.60.118 (119.0.6045.163) macOS 14.0   headless-new  selenium         10456
                                               real          selenium         10456
chrome  119                       Ubuntu 22.04 headless-new  selenium         10456
                                               xvfb          selenium         10456
                                  macOS 14.0   headless-new  selenium         10456
                                               real          selenium         10456
edge    119                       Ubuntu 22.04 headless-new  selenium         10456
                                               xvfb          selenium         10456
                                  macOS 14.0   headless-new  selenium         10456
                                               real          selenium         10456
firefox 119                       Ubuntu 22.04 headless      selenium         10456
                                               xvfb          selenium         10456
                                  macOS 14.0   headless      selenium         10456
                                               real          selenium         10456
safari  17.0                      macOS 14.0   real          selenium         10456

In [326]:
# Timeouts
display(df.groupby(["name", "version", "os", "headless_mode", "automation_mode"])["test_status"].value_counts().to_frame())

display(df.loc[df["test_status"] == 2])

count
name    version                   os           headless_mode automation_mode test_status       
brave   1.59.120 (118.0.5993.88)  Ubuntu 22.04 headless-new  selenium        0            10456
                                               xvfb          selenium        0            10456
        1.60.114 (119.0.6045.124) Ubuntu 22.04 headless-new  selenium        0            10456
                                               xvfb          selenium        0            10456
        1.60.118 (119.0.6045.163) macOS 14.0   headless-new  selenium        0            10456
                                               real          selenium        0            10456
chrome  119                       Ubuntu 22.04 headless-new  selenium        0            10456
                                               xvfb          selenium        0            10456
                                  macOS 14.0   headless-new  selenium        0            10456
                                               real          selenium        0            10456
edge    119                       Ubuntu 22.04 headless-new  selenium        0            10456
                                               xvfb          selenium        0            10455
                                                                             2                1
                                  macOS 14.0   headless-new  selenium        0            10455
                                                                             2                1
                                               real          selenium        0            10456
firefox 119                       Ubuntu 22.04 headless      selenium        0            10456
                                               xvfb          selenium        0            10456
                                  macOS 14.0   headless      selenium        0            10456
                                               real          selenium        0            10456
safari  17.0                      macOS 14.0   real          selenium        0            10388
                                                                             2               68

id        outcome_type outcome_value        test_name  \
51278    51391  <class 'NoneType'>          None  referrer_iframe   
51282    51395  <class 'NoneType'>          None  referrer_iframe   
51444    51399  <class 'NoneType'>          None  referrer_iframe   
52107    52079  <class 'NoneType'>          None  referrer_iframe   
52111    52083  <class 'NoneType'>          None  referrer_iframe   
...        ...                 ...           ...              ...   
62734    62678  <class 'NoneType'>          None  referrer_iframe   
62735    62679  <class 'NoneType'>          None  referrer_iframe   
62736    62680  <class 'NoneType'>          None  referrer_iframe   
153193  153240  <class 'NoneType'>          None      perfAPI_img   
155294  155232  <class 'NoneType'>          None  referrer_iframe   

        test_status    test_message test_stack org_scheme  \
51278             2  Test timed out       None       http   
51282             2  Test timed out       None       http   
51444             2  Test timed out       None       http   
52107             2  Test timed out       None       http   
52111             2  Test timed out       None       http   
...             ...             ...        ...        ...   
62734             2  Test timed out       None      https   
62735             2  Test timed out       None      https   
62736             2  Test timed out       None      https   
153193            2  Test timed out       None       http   
155294            2  Test timed out       None       http   

                           org_host resp_scheme  ... status_code label  \
51278   sub.headers.websec.saarland       https  ...         200    RP   
51282   sub.headers.websec.saarland       https  ...         200    RP   
51444   sub.headers.websec.saarland       https  ...         200    RP   
52107   sub.headers.websec.saarland       https  ...         200    RP   
52111   sub.headers.websec.saarland       https  ...         200    RP   
...                             ...         ...  ...         ...   ...   
62734   sub.headers.websec.saarland        http  ...         302    RP   
62735   sub.headers.websec.saarland       https  ...         302    RP   
62736   sub.headers.websec.saarland       https  ...         302    RP   
153193  sub.headers.websec.saarland       https  ...         200   TAO   
155294  sub.headers.websec.saarland       https  ...         302    RP   

        resp_type    name  version headless_mode            os  \
51278       basic  safari     17.0          real    macOS 14.0   
51282       basic  safari     17.0          real    macOS 14.0   
51444       basic  safari     17.0          real    macOS 14.0   
52107       basic  safari     17.0          real    macOS 14.0   
52111       basic  safari     17.0          real    macOS 14.0   
...           ...     ...      ...           ...           ...   
62734       basic  safari     17.0          real    macOS 14.0   
62735       basic  safari     17.0          real    macOS 14.0   
62736       basic  safari     17.0          real    macOS 14.0   
153193      basic    edge      119  headless-new    macOS 14.0   
155294      basic    edge      119          xvfb  Ubuntu 22.04   

       automation_mode add_info outcome_str  
51278         selenium                 None  
51282         selenium                 None  
51444         selenium                 None  
52107         selenium                 None  
52111         selenium                 None  
...                ...      ...         ...  
62734         selenium                 None  
62735         selenium                 None  
62736         selenium                 None  
153193        selenium                 None  
155294        selenium                 None  

[70 rows x 30 columns]

In [327]:
#TODO: easy way to redo timed out tests!

In [328]:
# Time taken!
df.groupby(["name", "version", "os", "headless_mode", "automation_mode", "org_scheme"])["created_at"].agg(["min", "max"]).apply(lambda x: x["max"] - x["min"], axis=1).to_frame()

0
name    version                   os           headless_mode automation_mode org_scheme                       
brave   1.59.120 (118.0.5993.88)  Ubuntu 22.04 headless-new  selenium        http       0 days 00:15:52.729060
                                                                             https      0 days 00:13:28.093020
                                               xvfb          selenium        http       0 days 00:15:53.539157
                                                                             https      0 days 00:13:33.120399
        1.60.114 (119.0.6045.124) Ubuntu 22.04 headless-new  selenium        http       0 days 00:15:50.330800
                                                                             https      0 days 00:13:29.664335
                                               xvfb          selenium        http       0 days 00:16:01.039583
                                                                             https      0 days 00:13:30.205549
        1.60.118 (119.0.6045.163) macOS 14.0   headless-new  selenium        http       0 days 00:16:00.851072
                                                                             https      0 days 00:13:17.676180
                                               real          selenium        http       0 days 00:16:47.333179
                                                                             https      0 days 00:13:51.071158
chrome  119                       Ubuntu 22.04 headless-new  selenium        http       0 days 00:15:32.276187
                                                                             https      0 days 00:13:11.597734
                                               xvfb          selenium        http       0 days 00:15:32.517923
                                                                             https      0 days 00:13:14.107244
                                  macOS 14.0   headless-new  selenium        http       0 days 00:10:32.869022
                                                                             https      0 days 00:13:15.386574
                                               real          selenium        http       0 days 00:11:05.393402
                                                                             https      0 days 00:13:43.572152
edge    119                       Ubuntu 22.04 headless-new  selenium        http       0 days 00:10:37.168142
                                                                             https      0 days 00:13:43.430620
                                               xvfb          selenium        http       0 days 00:10:29.766958
                                                                             https      0 days 00:13:35.611712
                                  macOS 14.0   headless-new  selenium        http       0 days 00:10:54.594401
                                                                             https      0 days 00:13:29.270315
                                               real          selenium        http       0 days 00:11:14.907553
                                                                             https      0 days 00:13:58.131905
firefox 119                       Ubuntu 22.04 headless      selenium        http       0 days 00:11:46.705398
                                                                             https      0 days 00:14:26.595717
                                               xvfb          selenium        http       0 days 00:12:03.017871
                                                                             https      0 days 00:14:44.644456
                                  macOS 14.0   headless      selenium        http       0 days 00:10:44.408936
                                                                             https      0 days 00:14:12.761382
                                               real          selenium        http       0 days 00:10:48.618109
  

## Response analysis

- We have 10456 unique tests: (test_name x relation_info) x org_scheme x (resp_scheme x resp_host) x response_id)
- They should all have 10 (num browsers in various settings) outcomes
    -  Due to the HTTPS Upgrade bug many have either 16 or 4 outcomes only (as the 6 http outcomes of Chrome + Brave instead resulted in HTTPS outcomes)
    -  Without org_scheme they should have 20; a couple tests have less than 20 (due to the Firefox errors and/or timeouts?)

In [329]:
# Unique tests
tests = {
    "Fetch": [4, 2, 8, 6], 
    "Framing": [9, 2, 8, 35], # 7+11+17
    "PP": [4, 2, 8, 24],
    "OAC": [3, 2, 8, 14],
    "TAO": [1, 2, 8, 13],
    "RP": [2, 2, 8, 16],
    "CSP-SCRIPT": [2, 2, 8, 17],
    "COEP": [2, 2, 8, 11],
    "CORP": [4, 2, 8, 11],
    "CSP-IMG": [1, 2, 8, 17],
    "HSTS": [2, 1, 2, 14],
    "COOP": [1, 2, 8, 11],
}
tests = pd.DataFrame.from_dict(tests, columns=["tests", "org_origins", "resp_origins", "responses"], orient="index")
tests["All tests"] = tests.prod(axis=1)
tests.loc["Sum"] = tests.sum(axis=0)
tests

tests  org_origins  resp_origins  responses  All tests
Fetch           4            2             8          6        384
Framing         9            2             8         35       5040
PP              4            2             8         24       1536
OAC             3            2             8         14        672
TAO             1            2             8         13        208
RP              2            2             8         16        512
CSP-SCRIPT      2            2             8         17        544
COEP            2            2             8         11        352
CORP            4            2             8         11        704
CSP-IMG         1            2             8         17        272
HSTS            2            1             2         14         56
COOP            1            2             8         11        176
Sum            35           23            90        189      10456

In [330]:
# SELECT count(*) as c2, count(DISTINCT browser_id), count(DISTINCT outcome_value) as c1, test_name, org_scheme, org_host, resp_scheme, resp_host, relation_info, response_id 
# FROM "Result" GROUP by test_name, org_scheme, org_host, resp_scheme, resp_host, relation_info, response_id order by c1 DESC, c2 DESC

res = df.groupby(["test_name", "org_scheme", "org_host", "resp_scheme", "resp_host", "relation_info", "response_id"])[["browser_id", "outcome_str"]].agg(["count", "nunique"])
display(res)
display(res.value_counts().to_frame())

browser_id  \
                                                                                                                                    count   
test_name             org_scheme org_host                    resp_scheme resp_host                   relation_info response_id              
accesswindow_direct   http       sub.headers.websec.saarland http        headers.webappsec.eu        direct        70                  19   
                                                                                                                   71                  19   
                                                                                                                   72                  19   
                                                                                                                   73                  19   
                                                                                                                   74                  19   
...                                                                                                                                   ...   
upgradeHSTS_subdomain http       sub.headers.websec.saarland https       sub.headers.websec.saarland subdomain     138                 19   
                                                                                                                   139                 19   
                                                                                                                   140                 19   
                                                                                                                   141                 19   
                                                                                                                   142                 19   

                                                                                                                                        \
                                                                                                                               nunique   
test_name             org_scheme org_host                    resp_scheme resp_host                   relation_info response_id           
accesswindow_direct   http       sub.headers.websec.saarland http        headers.webappsec.eu        direct        70               19   
                                                                                                                   71               19   
                                                                                                                   72               19   
                                                                                                                   73               19   
                                                                                                                   74               19   
...                                                                                                                                ...   
upgradeHSTS_subdomain http       sub.headers.websec.saarland https       sub.headers.websec.saarland subdomain     138              19   
                                                                                                                   139              19   
                                                                                                                   140              19   
                                                                                                                   141              19   
                                                                                                                   142              19   

                                                                                                                               outcome_str  \
                                                                             

count
(browser_id, count) (browser_id, nunique) (outcome_str, count) (outcome_str, nunique)       
19                  19                    19                   1                        8279
                                                               2                        1878
                                                               3                         294
                                                               4                           5

### Differences:
- COOP (accesswindow_direct):
    - [ ] Edge + Firefox: redirect first
    - [ ] Brave + Chrome: COOP first?
-  Framing:
    -  [ ] framing_iframe_sandbox: CSP-FA *; Firefox allows, all other browsers deny?
- PP not supported in Firefox
- OAC not supported in Firefox
    - [ ] Also random/caching behavior?
- PerfAPI:
    - [ ] Chrome/Firefox have different rules when to create an entry at all?
- COEP (CORP):
    - [ ] Firefox incorrect behavior for same-origin (applies same-site?)
- CORP:
    - [ ] same-origin: Edge does not check the scheme?
    - [x] Firefox applies CORP on object tags (even though it should not!)
    - [ ] Some other CORP and/or COEP strange stuff (Firefox)
- Referrer Policy: different default in Brave (unsafe-url is never allowed)
- Subresource-loading (CSP): differences due to HTTPS Upgrades
- Script Execution (CSP):
    - [ ] Safari sandbox bug?! (maybe related https://bugs.webkit.org/show_bug.cgi?id=223848 unsure though)
- Upgrade (HSTS):
    - [ ] Safari bug?

In [331]:
def get_uniques(df):
    browsers = f"{df['name'].unique().tolist()}"
    try:
        outcome = unique_outcomes[df.name[0]]
    except KeyError:
        outcome = {}
        #outcome = []
    outcome[browsers] = df.name[1]
    #outcome.append(f"{df.name[1]} ({browsers})")
    unique_outcomes[df.name[0]] = outcome

for grouping, group in df.groupby(["test_name", "relation_info"]):
    
    unique_outcomes = {}
    group.groupby(["name"])["outcome_str"].value_counts(normalize=True).reset_index().groupby(["outcome_str", "proportion"], group_keys=True).apply(get_uniques)
    d = pd.DataFrame.from_dict(unique_outcomes, orient="index")
    if d.shape[1] != 1:
        print(grouping)
        # display(unique_outcomes)
        display(group.groupby(["name"])["outcome_str"].value_counts(normalize=True).to_frame().unstack(0))
        for value, dic in unique_outcomes.items():
            values = [f"{v} ({k})" for k,v in dic.items()]
            # print(f"{value}: {values}")
        display(d)
        print()


('accesswindow_direct', 'direct')


proportion            \
name                                                    brave    chrome   
outcome_str                                                               
{'window.open.opener': 'null'}                       0.340909  0.261364   
{'window.open.opener': 'object "[object Window]"'}   0.659091  0.738636   

                                                                        \
name                                                    edge   firefox   
outcome_str                                                              
{'window.open.opener': 'null'}                      0.181818  0.181818   
{'window.open.opener': 'object "[object Window]"'}  0.818182  0.818182   

                                                              
name                                                  safari  
outcome_str                                                   
{'window.open.opener': 'null'}                      0.181818  
{'window.open.opener': 'object "[object Window]"'}  0.818182

['edge', 'firefox', 'safari']  \
{'window.open.opener': 'null'}                                           0.181818   
{'window.open.opener': 'object "[object Window]"'}                       0.818182   

                                                    ['chrome']  ['brave']  
{'window.open.opener': 'null'}                        0.261364   0.340909  
{'window.open.opener': 'object "[object Window]"'}    0.738636   0.659091


('framing_embed', 'direct')


proportion                                     
name                 brave    chrome      edge   firefox safari
outcome_str                                                    
message send      0.332143  0.332143  0.332143  0.332143    NaN
message timeout   0.667857  0.667857  0.667857  0.667857    1.0

['brave', 'chrome', 'edge', 'firefox']  ['safari']
message send                                   0.332143         NaN
message timeout                                0.667857         1.0


('framing_embed', 'nested')


proportion                                     
name                 brave    chrome      edge   firefox safari
outcome_str                                                    
message send      0.214286  0.214286  0.214286  0.214286    NaN
message timeout   0.785714  0.785714  0.785714  0.785714    1.0

['brave', 'chrome', 'edge', 'firefox']  ['safari']
message send                                   0.214286         NaN
message timeout                                0.785714         1.0


('framing_embed', 'sandbox')


proportion                                     
name                 brave    chrome      edge   firefox safari
outcome_str                                                    
message send      0.214286  0.214286  0.214286  0.214286    NaN
message timeout   0.785714  0.785714  0.785714  0.785714    1.0

['brave', 'chrome', 'edge', 'firefox']  ['safari']
message send                                   0.214286         NaN
message timeout                                0.785714         1.0


('framing_iframe', 'sandbox')


proportion                                        
name                 brave    chrome      edge   firefox    safari
outcome_str                                                       
message send      0.157143  0.157143  0.157143  0.171429  0.157143
message timeout   0.842857  0.842857  0.842857  0.828571  0.842857

['brave', 'chrome', 'edge', 'safari']  ['firefox']
message send                                  0.157143     0.171429
message timeout                               0.842857     0.828571


('framing_object', 'direct')


proportion                                     
name                 brave    chrome      edge   firefox safari
outcome_str                                                    
message send      0.332143  0.332143  0.332143  0.332143    NaN
message timeout   0.667857  0.667857  0.667857  0.667857    1.0

['brave', 'chrome', 'edge', 'firefox']  ['safari']
message send                                   0.332143         NaN
message timeout                                0.667857         1.0


('framing_object', 'nested')


proportion                                     
name                 brave    chrome      edge   firefox safari
outcome_str                                                    
message send      0.214286  0.214286  0.214286  0.214286    NaN
message timeout   0.785714  0.785714  0.785714  0.785714    1.0

['brave', 'chrome', 'edge', 'firefox']  ['safari']
message send                                   0.214286         NaN
message timeout                                0.785714         1.0


('framing_object', 'sandbox')


proportion                                     
name                 brave    chrome      edge   firefox safari
outcome_str                                                    
message send      0.214286  0.214286  0.214286  0.214286    NaN
message timeout   0.785714  0.785714  0.785714  0.785714    1.0

['brave', 'chrome', 'edge', 'firefox']  ['safari']
message send                                   0.214286         NaN
message timeout                                0.785714         1.0


('fullscreen_iframe', 'child')


proportion                                        
name                          brave    chrome      edge   firefox    safari
outcome_str                                                                
fullscreenEnabled: false   0.403646  0.403646  0.403646  0.343750  0.343750
fullscreenEnabled: true    0.054688  0.054688  0.054688  0.114583  0.114583
message timeout            0.541667  0.541667  0.541667  0.541667  0.541667

['firefox', 'safari']  ['brave', 'chrome', 'edge']  \
fullscreenEnabled: false               0.343750                     0.403646   
fullscreenEnabled: true                0.114583                     0.054688   
message timeout                             NaN                          NaN   

                          ['brave', 'chrome', 'edge', 'firefox', 'safari']  
fullscreenEnabled: false                                               NaN  
fullscreenEnabled: true                                                NaN  
message timeout                                                   0.541667


('fullscreen_iframe', 'child_allow')


proportion                                        
name                          brave    chrome      edge   firefox    safari
outcome_str                                                                
fullscreenEnabled: false   0.294271  0.294271  0.294271       NaN       NaN
fullscreenEnabled: true    0.164062  0.164062  0.164062  0.458333  0.458333
message timeout            0.541667  0.541667  0.541667  0.541667  0.541667

['brave', 'chrome', 'edge']  ['firefox', 'safari']  \
fullscreenEnabled: false                     0.294271                    NaN   
fullscreenEnabled: true                      0.164062               0.458333   
message timeout                                   NaN                    NaN   

                          ['brave', 'chrome', 'edge', 'firefox', 'safari']  
fullscreenEnabled: false                                               NaN  
fullscreenEnabled: true                                                NaN  
message timeout                                                   0.541667


('fullscreen_iframe', 'direct')


proportion                                    
name                          brave    chrome      edge firefox  safari
outcome_str                                                            
fullscreenEnabled: false   0.369792  0.369792  0.369792     NaN     NaN
fullscreenEnabled: true    0.317708  0.317708  0.317708  0.6875  0.6875
message timeout            0.312500  0.312500  0.312500  0.3125  0.3125

['brave', 'chrome', 'edge']  ['firefox', 'safari']  \
fullscreenEnabled: false                     0.369792                    NaN   
fullscreenEnabled: true                      0.317708                 0.6875   
message timeout                                   NaN                    NaN   

                          ['brave', 'chrome', 'edge', 'firefox', 'safari']  
fullscreenEnabled: false                                               NaN  
fullscreenEnabled: true                                                NaN  
message timeout                                                     0.3125


('imgloading_iframe', 'direct')


proportion                                        
name                 brave    chrome      edge   firefox    safari
outcome_str                                                       
error             0.194853  0.194853  0.194853  0.242647  0.242647
load              0.422794  0.422794  0.422794  0.375000  0.375000
message timeout   0.382353  0.382353  0.382353  0.382353  0.382353

['brave', 'chrome', 'edge']  ['firefox', 'safari']  \
error                               0.194853               0.242647   
load                                0.422794               0.375000   
message timeout                          NaN                    NaN   

                 ['brave', 'chrome', 'edge', 'firefox', 'safari']  
error                                                         NaN  
load                                                          NaN  
message timeout                                          0.382353


('oac_iframe', 'direct')


proportion                                \
name                                      brave    chrome      edge   firefox   
outcome_str                                                                     
message timeout                        0.357143  0.357143  0.357143  0.357143   
window.originAgentCluster: false       0.367560  0.245536  0.245536       NaN   
window.originAgentCluster: true        0.275298  0.397321  0.397321       NaN   
window.originAgentCluster: undefined        NaN       NaN       NaN  0.642857   

                                                
name                                    safari  
outcome_str                                     
message timeout                       0.357143  
window.originAgentCluster: false           NaN  
window.originAgentCluster: true            NaN  
window.originAgentCluster: undefined  0.642857

['brave', 'chrome', 'edge', 'firefox', 'safari']  \
message timeout                                                               0.357143   
window.originAgentCluster: false                                                   NaN   
window.originAgentCluster: true                                                    NaN   
window.originAgentCluster: undefined                                               NaN   

                                      ['chrome', 'edge']  ['brave']  \
message timeout                                      NaN        NaN   
window.originAgentCluster: false                0.245536   0.367560   
window.originAgentCluster: true                 0.397321   0.275298   
window.originAgentCluster: undefined                 NaN        NaN   

                                      ['firefox', 'safari']  
message timeout                                         NaN  
window.originAgentCluster: false                        NaN  
window.originAgentCluster: true                         NaN  
window.originAgentCluster: undefined               0.642857


('oac_iframe', 'sandbox')


proportion                                \
name                                      brave    chrome      edge   firefox   
outcome_str                                                                     
message timeout                        0.357143  0.357143  0.357143  0.357143   
window.originAgentCluster: true        0.642857  0.642857  0.642857       NaN   
window.originAgentCluster: undefined        NaN       NaN       NaN  0.642857   

                                                
name                                    safari  
outcome_str                                     
message timeout                       0.357143  
window.originAgentCluster: true            NaN  
window.originAgentCluster: undefined  0.642857

['brave', 'chrome', 'edge', 'firefox', 'safari']  \
message timeout                                                               0.357143   
window.originAgentCluster: true                                                    NaN   
window.originAgentCluster: undefined                                               NaN   

                                      ['brave', 'chrome', 'edge']  \
message timeout                                               NaN   
window.originAgentCluster: true                          0.642857   
window.originAgentCluster: undefined                          NaN   

                                      ['firefox', 'safari']  
message timeout                                         NaN  
window.originAgentCluster: true                         NaN  
window.originAgentCluster: undefined               0.642857


('oac_window.open', 'window.open')


proportion                                \
name                                      brave    chrome      edge   firefox   
outcome_str                                                                     
message timeout                        0.142857  0.142857  0.142857  0.142857   
window.originAgentCluster: false       0.340774  0.285714  0.459821       NaN   
window.originAgentCluster: true        0.516369  0.571429  0.397321       NaN   
window.originAgentCluster: undefined        NaN       NaN       NaN  0.857143   

                                                
name                                    safari  
outcome_str                                     
message timeout                       0.142857  
window.originAgentCluster: false           NaN  
window.originAgentCluster: true            NaN  
window.originAgentCluster: undefined  0.857143

['brave', 'chrome', 'edge', 'firefox', 'safari']  \
message timeout                                                               0.142857   
window.originAgentCluster: false                                                   NaN   
window.originAgentCluster: true                                                    NaN   
window.originAgentCluster: undefined                                               NaN   

                                      ['chrome']  ['brave']  ['edge']  \
message timeout                              NaN        NaN       NaN   
window.originAgentCluster: false        0.285714   0.340774  0.459821   
window.originAgentCluster: true         0.571429   0.516369  0.397321   
window.originAgentCluster: undefined         NaN        NaN       NaN   

                                      ['firefox', 'safari']  
message timeout                                         NaN  
window.originAgentCluster: false                        NaN  
window.originAgentCluster: true                         NaN  
window.originAgentCluster: undefined               0.857143


('perfAPI_img', 'direct')


proportion                                \
name                              brave    chrome      edge   firefox   
outcome_str                                                             
No performance entry           0.250000  0.250000  0.250000       NaN   
None                                NaN       NaN  0.001202       NaN   
{'requestStart != 0': False}   0.442308  0.442308  0.442308  0.581731   
{'requestStart != 0': True}    0.307692  0.307692  0.306490  0.418269   

                                        
name                            safari  
outcome_str                             
No performance entry               NaN  
None                               NaN  
{'requestStart != 0': False}  0.581731  
{'requestStart != 0': True}   0.418269

['brave', 'chrome', 'edge']  ['edge']  \
No performance entry                             0.250000       NaN   
{'requestStart != 0': False}                     0.442308       NaN   
None                                                  NaN  0.001202   
{'requestStart != 0': True}                           NaN  0.306490   

                              ['firefox', 'safari']  ['brave', 'chrome']  
No performance entry                            NaN                  NaN  
{'requestStart != 0': False}               0.581731                  NaN  
None                                            NaN                  NaN  
{'requestStart != 0': True}                0.418269             0.307692


('referrer_iframe', 'iframe')


proportion            \
name                                                    brave    chrome   
outcome_str                                                               
None                                                      NaN       NaN   
document.referrer:                                   0.054688  0.054688   
document.referrer: http://headers.webappsec.eu/      0.044922  0.035156   
document.referrer: http://headers.webappsec.eu/...   0.001953  0.011719   
document.referrer: http://headers.websec.saarland/   0.044922  0.035156   
document.referrer: http://headers.websec.saarla...   0.001953  0.011719   
document.referrer: http://sub.headers.websec.sa...   0.011719  0.011719   
document.referrer: http://sub.headers.websec.sa...   0.039062  0.039062   
document.referrer: http://sub.sub.headers.webse...   0.044922  0.035156   
document.referrer: http://sub.sub.headers.webse...   0.001953  0.011719   
document.referrer: https://headers.webappsec.eu/     0.044922  0.035156   
document.referrer: https://headers.webappsec.eu...   0.001953  0.011719   
document.referrer: https://headers.websec.saarl...   0.044922  0.035156   
document.referrer: https://headers.websec.saarl...   0.001953  0.011719   
document.referrer: https://sub.headers.websec.s...   0.011719  0.011719   
document.referrer: https://sub.headers.websec.s...   0.039062  0.039062   
document.referrer: https://sub.sub.headers.webs...   0.044922  0.035156   
document.referrer: https://sub.sub.headers.webs...   0.001953  0.011719   
message timeout                                      0.562500  0.562500   

                                                                        \
name                                                    edge   firefox   
outcome_str                                                              
None                                                     NaN       NaN   
document.referrer:                                  0.054688  0.054688   
document.referrer: http://headers.webappsec.eu/     0.035156  0.046875   
document.referrer: http://headers.webappsec.eu/...  0.011719       NaN   
document.referrer: http://headers.websec.saarland/  0.035156  0.035156   
document.referrer: http://headers.websec.saarla...  0.011719  0.011719   
document.referrer: http://sub.headers.websec.sa...  0.011719  0.011719   
document.referrer: http://sub.headers.websec.sa...  0.039062  0.039062   
document.referrer: http://sub.sub.headers.webse...  0.035156  0.035156   
document.referrer: http://sub.sub.headers.webse...  0.011719  0.011719   
document.referrer: https://headers.webappsec.eu/    0.035156  0.046875   
document.referrer: https://headers.webappsec.eu...  0.011719       NaN   
document.referrer: https://headers.websec.saarl...  0.035156  0.035156   
document.referrer: https://headers.websec.saarl...  0.011719  0.011719   
document.referrer: https://sub.headers.websec.s...  0.011719  0.011719   
document.referrer: https://sub.headers.websec.s...  0.039062  0.039062   
document.referrer: https://sub.sub.headers.webs...  0.035156  0.035156   
document.referrer: https://sub.sub.headers.webs...  0.011719  0.011719   
message timeout                                     0.562500  0.562500   

                                                              
name                                                  safari  
outcome_str                                                   
None                                                0.160156  
document.referrer:                                  0.054688  
document.referrer: http://headers.webappsec.eu/     0.046875  
document.referrer: http://headers.webappsec.eu/...       NaN  
document.referrer: http://headers.websec.saarland/  0.035156  
document.referrer: http://headers.websec.saarla...  0.011719  
document.referrer: http://sub.headers.websec.sa...  0.011719  
document.referrer: http://sub.headers.websec.sa...  0.039062  
document.referrer: http://sub.sub.headers.webse...  0.035156  
document.referr

['safari']  \
None                                                  0.160156   
message timeout                                       0.402344   
document.referrer:                                         NaN   
document.referrer: http://sub.headers.websec.sa...         NaN   
document.referrer: http://sub.headers.websec.sa...         NaN   
document.referrer: https://sub.headers.websec.s...         NaN   
document.referrer: https://sub.headers.websec.s...         NaN   
document.referrer: http://headers.webappsec.eu/            NaN   
document.referrer: http://headers.webappsec.eu/...         NaN   
document.referrer: https://headers.webappsec.eu/           NaN   
document.referrer: https://headers.webappsec.eu...         NaN   
document.referrer: http://headers.websec.saarland/         NaN   
document.referrer: http://headers.websec.saarla...         NaN   
document.referrer: http://sub.sub.headers.webse...         NaN   
document.referrer: http://sub.sub.headers.webse...         NaN   
document.referrer: https://headers.websec.saarl...         NaN   
document.referrer: https://headers.websec.saarl...         NaN   
document.referrer: https://sub.sub.headers.webs...         NaN   
document.referrer: https://sub.sub.headers.webs...         NaN   

                                                    ['brave', 'chrome', 'edge', 'firefox', 'safari']  \
None                                                                                             NaN   
message timeout                                                                                  NaN   
document.referrer:                                                                          0.054688   
document.referrer: http://sub.headers.websec.sa...                                          0.011719   
document.referrer: http://sub.headers.websec.sa...                                          0.039062   
document.referrer: https://sub.headers.websec.s...                                          0.011719   
document.referrer: https://sub.headers.websec.s...                                          0.039062   
document.referrer: http://headers.webappsec.eu/                                                  NaN   
document.referrer: http://headers.webappsec.eu/...                                               NaN   
document.referrer: https://headers.webappsec.eu/                                                 NaN   
document.referrer: https://headers.webappsec.eu...                                               NaN   
document.referrer: http://headers.websec.saarland/                                               NaN   
document.referrer: http://headers.websec.saarla...                                               NaN   
document.referrer: http://sub.sub.headers.webse...                                               NaN   
document.referrer: http://sub.sub.headers.webse...                                               NaN   
document.referrer: https://headers.websec.saarl...                                               NaN   
document.referrer: https://headers.websec.saarl...                                               NaN   
document.referrer: https://sub.sub.headers.webs...                                               NaN   
document.referrer: https://sub.sub.headers.webs...                                               NaN   

                                                    ['chrome', 'edge']  \
None                                                               NaN   
message timeout                                                    NaN   
document.referrer:                                                 NaN   
document.referrer: http://sub.headers.websec.sa...                 NaN   
document.referrer: http://sub.headers.websec.sa...                 NaN   
document.referrer: https://sub.headers.websec.s...                 NaN   
document.referrer: https://sub.headers.websec.s...                 NaN   
document.referrer: http://headers.webappsec.eu/           


('referrer_iframe', 'window.open')


proportion            \
name                                                    brave    chrome   
outcome_str                                                               
None                                                      NaN       NaN   
document.referrer:                                   0.427083  0.289062   
document.referrer: http://headers.webappsec.eu/           NaN  0.035156   
document.referrer: http://headers.webappsec.eu/...        NaN  0.005859   
document.referrer: http://headers.webappsec.eu/...        NaN  0.005859   
document.referrer: http://headers.websec.saarland/        NaN  0.035156   
document.referrer: http://headers.websec.saarla...        NaN  0.005859   
document.referrer: http://headers.websec.saarla...        NaN  0.005859   
document.referrer: http://sub.headers.websec.sa...   0.042318  0.041016   
document.referrer: http://sub.headers.websec.sa...   0.004557  0.025391   
document.referrer: http://sub.headers.websec.sa...        NaN  0.005859   
document.referrer: http://sub.sub.headers.webse...        NaN  0.035156   
document.referrer: http://sub.sub.headers.webse...        NaN  0.005859   
document.referrer: http://sub.sub.headers.webse...        NaN  0.005859   
document.referrer: https://headers.webappsec.eu/     0.092448  0.064453   
document.referrer: https://headers.webappsec.eu...   0.002604  0.011719   
document.referrer: https://headers.webappsec.eu...   0.003906  0.017578   
document.referrer: https://headers.websec.saarl...   0.092448  0.064453   
document.referrer: https://headers.websec.saarl...   0.002604  0.011719   
document.referrer: https://headers.websec.saarl...   0.003906  0.017578   
document.referrer: https://sub.headers.websec.s...   0.024740  0.025391   
document.referrer: https://sub.headers.websec.s...   0.001302  0.007812   
document.referrer: https://sub.headers.websec.s...   0.078125  0.058594   
document.referrer: https://sub.sub.headers.webs...   0.092448  0.064453   
document.referrer: https://sub.sub.headers.webs...   0.002604  0.011719   
document.referrer: https://sub.sub.headers.webs...   0.003906  0.017578   
message timeout                                      0.125000  0.125000   

                                                                        \
name                                                    edge   firefox   
outcome_str                                                              
None                                                0.000977       NaN   
document.referrer:                                  0.242188  0.242188   
document.referrer: http://headers.webappsec.eu/     0.070312  0.070312   
document.referrer: http://headers.webappsec.eu/...  0.011719  0.011719   
document.referrer: http://headers.webappsec.eu/...  0.011719  0.011719   
document.referrer: http://headers.websec.saarland/  0.070312  0.070312   
document.referrer: http://headers.websec.saarla...  0.011719  0.011719   
document.referrer: http://headers.websec.saarla...  0.011719  0.011719   
document.referrer: http://sub.headers.websec.sa...  0.046875  0.046875   
document.referrer: http://sub.headers.websec.sa...  0.039062  0.039062   
document.referrer: http://sub.headers.websec.sa...  0.011719  0.011719   
document.referrer: http://sub.sub.headers.webse...  0.070312  0.070312   
document.referrer: http://sub.sub.headers.webse...  0.011719  0.011719   
document.referrer: http://sub.sub.headers.webse...  0.011719  0.011719   
document.referrer: https://headers.webappsec.eu/    0.042969  0.042969   
document.referrer: https://headers.webappsec.eu...  0.007812  0.007812   
document.referrer: https://headers.webappsec.eu...  0.011719  0.011719   
document.referrer: https://headers.websec.saarl...  0.042969  0.042969   
document.referrer: https://headers.websec.saarl...  0.007812  0.007812   
document.referrer: https://headers.websec.saarl...  0.011719  0.011719   
document.referrer: https://sub.headers.websec.s...  0.019531  0.019531   
document.referrer: https

['edge']  ['safari']  \
None                                                0.000977    0.105469   
message timeout                                     0.124023    0.019531   
document.referrer: http://headers.webappsec.eu/          NaN    0.093750   
document.referrer: https://headers.webappsec.eu/         NaN    0.062500   
document.referrer:                                       NaN         NaN   
document.referrer: http://headers.websec.saarland/       NaN         NaN   
document.referrer: http://headers.websec.saarla...       NaN         NaN   
document.referrer: http://headers.websec.saarla...       NaN         NaN   
document.referrer: http://sub.headers.websec.sa...       NaN         NaN   
document.referrer: http://sub.headers.websec.sa...       NaN         NaN   
document.referrer: http://sub.headers.websec.sa...       NaN         NaN   
document.referrer: http://sub.sub.headers.webse...       NaN         NaN   
document.referrer: http://sub.sub.headers.webse...       NaN         NaN   
document.referrer: http://sub.sub.headers.webse...       NaN         NaN   
document.referrer: https://headers.websec.saarl...       NaN         NaN   
document.referrer: https://headers.websec.saarl...       NaN         NaN   
document.referrer: https://headers.websec.saarl...       NaN         NaN   
document.referrer: https://sub.headers.websec.s...       NaN         NaN   
document.referrer: https://sub.headers.websec.s...       NaN         NaN   
document.referrer: https://sub.sub.headers.webs...       NaN         NaN   
document.referrer: https://sub.sub.headers.webs...       NaN         NaN   
document.referrer: https://sub.sub.headers.webs...       NaN         NaN   
document.referrer: http://headers.webappsec.eu/...       NaN         NaN   
document.referrer: http://headers.webappsec.eu/...       NaN         NaN   
document.referrer: https://headers.webappsec.eu...       NaN         NaN   
document.referrer: https://headers.webappsec.eu...       NaN         NaN   
document.referrer: https://sub.headers.websec.s...       NaN         NaN   

                                                    ['edge', 'firefox', 'safari']  \
None                                                                          NaN   
message timeout                                                               NaN   
document.referrer: http://headers.webappsec.eu/                               NaN   
document.referrer: https://headers.webappsec.eu/                              NaN   
document.referrer:                                                       0.242188   
document.referrer: http://headers.websec.saarland/                       0.070312   
document.referrer: http://headers.websec.saarla...                       0.011719   
document.referrer: http://headers.websec.saarla...                       0.011719   
document.referrer: http://sub.headers.websec.sa...                       0.046875   
document.referrer: http://sub.headers.websec.sa...                       0.039062   
document.referrer: http://sub.headers.websec.sa...                       0.011719   
document.referrer: http://sub.sub.headers.webse...                       0.070312   
document.referrer: http://sub.sub.headers.webse...                       0.011719   
document.referrer: http://sub.sub.headers.webse...                       0.011719   
document.referrer: https://headers.websec.saarl...                       0.042969   
document.referrer: https://headers.websec.saarl...                       0.007812   
document.referrer: https://headers.websec.saarl...                       0.011719   
document.referrer: https://sub.headers.websec.s...                       0.019531   
document.referrer: https://sub.headers.websec.s...                       0.039062   
document.referrer: https://sub.sub.headers.webs...                       0.042969   
document.referrer: https://sub.sub.headers.webs...                       0.007812   
document.referrer: https://sub.sub.headers.webs...   


('script_execution_iframe', 'sandbox')


proportion                                        
name                 brave    chrome      edge   firefox    safari
outcome_str                                                       
message send      0.496324  0.496324  0.496324  0.496324  0.485294
message timeout   0.503676  0.503676  0.503676  0.503676  0.514706

['safari']  ['brave', 'chrome', 'edge', 'firefox']
message send       0.485294                                0.496324
message timeout    0.514706                                0.503676


('subresourceloadingCOEP_img', 'direct')


proportion            \
name                                                    brave    chrome   
outcome_str                                                               
message timeout                                      0.386364  0.386364   
{'image-events': {'swag.jpg': 'error', 'swag-sa...   0.005682  0.005682   
{'image-events': {'swag.jpg': 'error', 'swag-sa...   0.011364  0.011364   
{'image-events': {'swag.jpg': 'load', 'swag-sam...   0.147727  0.147727   
{'image-events': {'swag.jpg': 'load', 'swag-sam...   0.295455  0.295455   
{'image-events': {'swag.jpg': 'load', 'swag-sam...   0.153409  0.153409   

                                                                        \
name                                                    edge   firefox   
outcome_str                                                              
message timeout                                     0.386364  0.386364   
{'image-events': {'swag.jpg': 'error', 'swag-sa...  0.005682  0.005682   
{'image-events': {'swag.jpg': 'error', 'swag-sa...  0.011364       NaN   
{'image-events': {'swag.jpg': 'load', 'swag-sam...  0.147727  0.147727   
{'image-events': {'swag.jpg': 'load', 'swag-sam...  0.295455  0.183239   
{'image-events': {'swag.jpg': 'load', 'swag-sam...  0.153409  0.276989   

                                                              
name                                                  safari  
outcome_str                                                   
message timeout                                     0.386364  
{'image-events': {'swag.jpg': 'error', 'swag-sa...  0.011364  
{'image-events': {'swag.jpg': 'error', 'swag-sa...  0.028409  
{'image-events': {'swag.jpg': 'load', 'swag-sam...  0.142045  
{'image-events': {'swag.jpg': 'load', 'swag-sam...  0.329545  
{'image-events': {'swag.jpg': 'load', 'swag-sam...  0.102273

['brave', 'chrome', 'edge', 'firefox', 'safari']  \
message timeout                                                                             0.386364   
{'image-events': {'swag.jpg': 'error', 'swag-sa...                                               NaN   
{'image-events': {'swag.jpg': 'load', 'swag-sam...                                               NaN   
{'image-events': {'swag.jpg': 'error', 'swag-sa...                                               NaN   
{'image-events': {'swag.jpg': 'load', 'swag-sam...                                               NaN   
{'image-events': {'swag.jpg': 'load', 'swag-sam...                                               NaN   

                                                    ['brave', 'chrome', 'edge', 'firefox']  \
message timeout                                                                        NaN   
{'image-events': {'swag.jpg': 'error', 'swag-sa...                                0.005682   
{'image-events': {'swag.jpg': 'load', 'swag-sam...                                0.147727   
{'image-events': {'swag.jpg': 'error', 'swag-sa...                                     NaN   
{'image-events': {'swag.jpg': 'load', 'swag-sam...                                     NaN   
{'image-events': {'swag.jpg': 'load', 'swag-sam...                                     NaN   

                                                    ['safari']  \
message timeout                                            NaN   
{'image-events': {'swag.jpg': 'error', 'swag-sa...    0.011364   
{'image-events': {'swag.jpg': 'load', 'swag-sam...    0.142045   
{'image-events': {'swag.jpg': 'error', 'swag-sa...    0.028409   
{'image-events': {'swag.jpg': 'load', 'swag-sam...    0.329545   
{'image-events': {'swag.jpg': 'load', 'swag-sam...    0.102273   

                                                    ['brave', 'chrome', 'edge']  \
message timeout                                                             NaN   
{'image-events': {'swag.jpg': 'error', 'swag-sa...                          NaN   
{'image-events': {'swag.jpg': 'load', 'swag-sam...                          NaN   
{'image-events': {'swag.jpg': 'error', 'swag-sa...                     0.011364   
{'image-events': {'swag.jpg': 'load', 'swag-sam...                     0.295455   
{'image-events': {'swag.jpg': 'load', 'swag-sam...                     0.153409   

                                                    ['firefox']  
message timeout                                             NaN  
{'image-events': {'swag.jpg': 'error', 'swag-sa...          NaN  
{'image-events': {'swag.jpg': 'load', 'swag-sam...          NaN  
{'image-events': {'swag.jpg': 'error', 'swag-sa...          NaN  
{'image-events': {'swag.jpg': 'load', 'swag-sam...     0.183239  
{'image-events': {'swag.jpg': 'load', 'swag-sam...     0.276989


('subresourceloadingCOEP_img', 'sandbox')


proportion            \
name                                                    brave    chrome   
outcome_str                                                               
message timeout                                      0.386364  0.386364   
{'image-events': {'swag.jpg': 'error', 'swag-sa...   0.022727  0.022727   
{'image-events': {'swag.jpg': 'load', 'swag-sam...   0.590909  0.590909   

                                                                        \
name                                                    edge   firefox   
outcome_str                                                              
message timeout                                     0.386364  0.386364   
{'image-events': {'swag.jpg': 'error', 'swag-sa...  0.022727  0.022727   
{'image-events': {'swag.jpg': 'load', 'swag-sam...  0.590909  0.590909   

                                                              
name                                                  safari  
outcome_str                                                   
message timeout                                     0.386364  
{'image-events': {'swag.jpg': 'error', 'swag-sa...  0.045455  
{'image-events': {'swag.jpg': 'load', 'swag-sam...  0.568182

['brave', 'chrome', 'edge', 'firefox', 'safari']  \
message timeout                                                                             0.386364   
{'image-events': {'swag.jpg': 'error', 'swag-sa...                                               NaN   
{'image-events': {'swag.jpg': 'load', 'swag-sam...                                               NaN   

                                                    ['brave', 'chrome', 'edge', 'firefox']  \
message timeout                                                                        NaN   
{'image-events': {'swag.jpg': 'error', 'swag-sa...                                0.022727   
{'image-events': {'swag.jpg': 'load', 'swag-sam...                                0.590909   

                                                    ['safari']  
message timeout                                            NaN  
{'image-events': {'swag.jpg': 'error', 'swag-sa...    0.045455  
{'image-events': {'swag.jpg': 'load', 'swag-sam...    0.568182


('subresourceloadingCORP_img', 'direct')


proportion                                        
name             brave    chrome      edge   firefox    safari
outcome_str                                                   
error         0.295455  0.295455  0.295455  0.301136  0.301136
load          0.704545  0.704545  0.704545  0.698864  0.698864

['brave', 'chrome', 'edge']  ['firefox', 'safari']
error                     0.295455               0.301136
load                      0.704545               0.698864


('subresourceloadingCORP_object', 'direct')


proportion                              
name                         brave chrome  edge   firefox safari
outcome_str                                                     
error                         0.25   0.25  0.25  0.431818    NaN
load                          0.75   0.75  0.75  0.568182    NaN
timeout: no event fired        NaN    NaN   NaN       NaN    1.0

['brave', 'chrome', 'edge']  ['firefox']  ['safari']
error                                           0.25     0.431818         NaN
load                                            0.75     0.568182         NaN
timeout: no event fired                          NaN          NaN         1.0


('upgradeHSTS_direct', 'direct')


proportion                                \
name                                brave    chrome      edge   firefox   
outcome_str                                                               
{'response.redirected': False}   0.821429  0.821429  0.821429  0.821429   
{'response.redirected': True}    0.178571  0.178571  0.178571  0.178571   

                                          
name                              safari  
outcome_str                               
{'response.redirected': False}  0.785714  
{'response.redirected': True}   0.214286

['safari']  \
{'response.redirected': False}    0.785714   
{'response.redirected': True}     0.214286   

                                ['brave', 'chrome', 'edge', 'firefox']  
{'response.redirected': False}                                0.821429  
{'response.redirected': True}                                 0.178571

In [334]:
def make_clickable(url):
    # Clickable links for debugging; set browser_id=2 (unknown) to not count them in our real data collection
    url = re.sub(r"browser_id=(\d+)", "browser_id=2", url)
    return f'<a href="{url}" target="_blank">{url}</a>'

test_name_seen = set()
for test_name, org_scheme, org_host, resp_scheme, resp_host, relation_info, response_id in res.loc[res[("outcome_str", "nunique")] != 1].index:
    # Only look at each feature group once
    if test_name in test_name_seen:
        continue
    test_name_seen.add(test_name)
    print(test_name)
    rows = df.loc[(df["org_scheme"] == org_scheme) & (df["test_name"] == test_name) & (df["resp_scheme"] == resp_scheme) & (df["resp_host"] == resp_host) & (df["relation_info"] == relation_info) & (df["response_id"] == response_id)]
    with pd.option_context("display.max_colwidth", 200):
        #display(rows[["outcome_str", "name", "headless_mode", "os", "version"]].value_counts().to_frame())
        display(rows[["outcome_str", "name"]].value_counts().to_frame())
        disp = rows.drop_duplicates(subset="outcome_str")[["browser_id", "headless_mode", "name", "outcome_str", "raw_header", "test_name", "org_scheme", "org_host", "resp_scheme", "relation_info", "resp_host", "full_url"]]
        disp = disp.style.format({'full_url': make_clickable})
        display(disp)
        
    

accesswindow_direct


count
outcome_str                                        name          
{'window.open.opener': 'null'}                     brave        6
{'window.open.opener': 'object "[object Window]"'} edge         4
                                                   firefox      4
{'window.open.opener': 'null'}                     chrome       2
{'window.open.opener': 'object "[object Window]"'} chrome       2
                                                   safari       1

framing_embed


count
outcome_str     name          
message send    brave        6
                chrome       4
                edge         4
                firefox      4
message timeout safari       1

framing_iframe


count
outcome_str     name          
message timeout brave        6
message send    firefox      4
message timeout chrome       4
                edge         4
                safari       1

framing_object


count
outcome_str     name          
message send    brave        6
                chrome       4
                edge         4
                firefox      4
message timeout safari       1

fullscreen_iframe


count
outcome_str              name          
fullscreenEnabled: false brave        6
                         chrome       4
                         edge         4
fullscreenEnabled: true  firefox      4
                         safari       1

imgloading_iframe


count
outcome_str name          
load        brave        6
error       firefox      4
load        chrome       4
            edge         4
error       safari       1

oac_iframe


count
outcome_str                          name          
window.originAgentCluster: false     brave        6
                                     chrome       4
                                     edge         4
window.originAgentCluster: undefined firefox      4
                                     safari       1

oac_window.open


count
outcome_str                          name          
window.originAgentCluster: false     edge         4
window.originAgentCluster: undefined firefox      4
window.originAgentCluster: true      brave        4
window.originAgentCluster: false     chrome       2
                                     brave        2
window.originAgentCluster: true      chrome       2
window.originAgentCluster: undefined safari       1

perfAPI_img


count
outcome_str                 name          
{'requestStart != 0': True} brave        6
                            chrome       4
                            firefox      4
                            edge         3
None                        edge         1
{'requestStart != 0': True} safari       1

referrer_iframe


count
outcome_str                                                                                                                                                                                    name          
document.referrer: http://headers.webappsec.eu/                                                                                                                                                brave        5
                                                                                                                                                                                               firefox      4
document.referrer: http://headers.webappsec.eu/_hp/server/responses.py?feature_group=rp&resp_id=<resp_id>&count=<count>&nest=1&origin=http://sub.headers.websec.saarland&element=iframe&resp=1 edge         4
                                                                                                                                                                                               chrome       4
                                                                                                                                                                                               brave        1
document.referrer: http://headers.webappsec.eu/                                                                                                                                                safari       1

script_execution_iframe


count
outcome_str     name          
message timeout brave        6
                edge         4
                chrome       4
                firefox      4
message send    safari       1

subresourceloadingCOEP_img


count
outcome_str                                                                                                                                                               name          
{'image-events': {'swag.jpg': 'load', 'swag-same-site.jpg': 'load', 'swag-cross-site.jpg': 'load', 'swag-same-origin.jpg': 'error'}, 'window.crossOriginIsolated': False} brave        6
                                                                                                                                                                          chrome       4
                                                                                                                                                                          edge         4
{'image-events': {'swag.jpg': 'load', 'swag-same-site.jpg': 'load', 'swag-cross-site.jpg': 'load', 'swag-same-origin.jpg': 'load'}, 'window.crossOriginIsolated': False}  firefox      3
{'image-events': {'swag.jpg': 'load', 'swag-same-site.jpg': 'load', 'swag-cross-site.jpg': 'load', 'swag-same-origin.jpg': 'error'}, 'window.crossOriginIsolated': False} firefox      1
                                                                                                                                                                          safari       1

subresourceloadingCORP_img


count
outcome_str name          
load        brave        6
error       firefox      4
load        chrome       4
            edge         4
error       safari       1

subresourceloadingCORP_object


count
outcome_str             name          
load                    brave        6
                        chrome       4
                        edge         4
                        firefox      4
timeout: no event fired safari       1

upgradeHSTS_direct


count
outcome_str                    name          
{'response.redirected': False} brave        6
                               chrome       4
                               edge         4
                               firefox      4
{'response.redirected': True}  safari       1

# Crash analysis

### Exceptions thrown
- Fixed:
    - Only Firefox; both headless + Xvfb
        - OAC + RP
        - Both on HTTP and HTTPS
        - Response_ids: 185, 186, 341, 342; these are the ones that have status code 302!
        - Reason seems to be that Firefox switches to another tab in these cases and thus the wait for crashes. Solution: forcefully switch back to the original tab before waiting!
    - Only Brave: both headless + Xvfb
        - OAC + RP
        - Both on HTTP and HTTPS
        - Reason seems to be the tab switching, if deactivated no issue occurs

In [147]:
with open("/home/ubuntu/wpt/_hp/tools/crawler/desktop-selenium-2023-11-20 16:09:24.649008.log", "r") as f:
    crashes = []
    for l in f.readlines():
        if "649008: http" in l:
            match = re.match(".*(http|https)://.*browser_id=(\d+)&label=(.*)&first_id=(\d+).*", l)
            crashes.append([match[1], match[2], match[3], match[4]])

In [148]:
crashes = pd.DataFrame(crashes, columns=["scheme", "browser_id", "label", "response_id"]).drop_duplicates()
crashes

scheme browser_id label response_id
0    http         59    RP         172
2    http         59    RP         173
4   https         59   OAC         339
6   https         59    RP         181
8    http         75    RP         171
10   http         75    RP         172
12  https         75   OAC         330
14  https         75   OAC         338
16  https         75    RP         180
18  https         74   OAC         337
20  https         74    RP         175